sway-lsp/src/server.rs

```diff
@@ -486,237 +514,448 @@ mod tests {
+                            "start": {
+                                "line": 3,
+                                "character": 4
+                            },
+                            "end": {
+                                "line": 3,
+                                "character": 4
+                            }
+                        },
+                        "rangeLength": 0,
+                        "text": "let x = 0.0;",
+                    }
+                ]
+            });
 
-        // send "shutdown" request
-        let _ = shutdown_request(&mut service).await;
+            let _ = did_change_request(&mut service, params).await;
 
-        // send "exit" request
-        exit_notification(&mut service).await;
+            shutdown_and_exit(&mut service).await;
+        });
+        rt.shutdown_background();
     }
 
-    //#[tokio::test]
-    #[allow(dead_code)]
-    async fn show_ast() {
-        let (mut service, mut messages) =
-            LspService::build(|client| Backend::new(client, config()))
+    #[test]
+    #[serial]
+    fn show_ast() {
+        let rt = tokio::runtime::Runtime::new().unwrap();
+        rt.block_on(async {
+            let (mut service, _) = LspService::build(|client| Backend::new(client, config()))
                 .custom_method("sway/show_ast", Backend::show_ast)
                 .finish();
 
-        // send "initialize" request
-        let _ = initialize_request(&mut service).await;
+            let uri = init_and_open(&mut service, e2e_test_dir()).await;
 
-        // send "initialized" notification
-        initialized_notification(&mut service).await;
+            // send "sway/show_typed_ast" request
+            let _ = show_ast_request(&mut service, &uri).await;
 
-        // ignore the "window/logMessage" notification: "Initializing the Sway Language Server"
-        messages.next().await.unwrap();
+            shutdown_and_exit(&mut service).await;
+        });
+        rt.shutdown_background();
+    }
+
+    // This macro allows us to spin up a server / client for testing
+    // It initializes and performs the necessary handshake and then loads
+    // the sway example that was passed into `example_dir`.
+    // It then runs the specific capability to test before gracefully shutting down.
+    // The capability argument is an async function.
+    macro_rules! test_lsp_capability {
+        ($example_dir:expr, $capability:expr) => {{
+            let rt = tokio::runtime::Runtime::new().unwrap();
+            rt.block_on(async {
+                let (mut service, _) = LspService::new(|client| Backend::new(client, config()));
+                let uri = init_and_open(&mut service, $example_dir).await;
+                // Call the specific LSP capability function that was passed in.
+                let _ = $capability(&mut service, &uri).await;
+
+                shutdown_and_exit(&mut service).await;
+            });
+            rt.shutdown_background();
+        }};
+    }
```

Comment:
We had a good go at this @Centril , but it would require being able to pass an `async` function as an argument to the higher order `test_lsp_capability` function, which is a bit tricky at the moment as currently Rust only supports `impl Trait` in function argument and return type position (i.e. not in trait position, which would be required for the function argument). But feel free to have a go at it!

As a result, I suggested going with a small well-documented `macro_rules!` instead seeing as it's not publicly exposed and only used by the tests.

Reply:
As @mitchmindtree stated above, this is as far as we got. Would love to know if you could find a solution though. 

```rust 
pub async fn test_lsp_cap<F, Fut>(example_dir: PathBuf, capability: F)
    where
        F: Fn(&mut LspService<Backend>, &Url) -> Fut,
        Fut: std::future::Future<Output = Request> + 'static + Send,
{
    let (mut service, _) = LspService::new(|client| Backend::new(client, config()));
    let uri = init_and_open(&mut service, example_dir).await;
    // Call the specific LSP capability function that was passed in.
    let _ = capability(&mut service, &uri).await;
    shutdown_and_exit(&mut service).await;
}

#[tokio::test]
async fn semantic_tokens() {
    test_lsp_cap(doc_comments_dir(), semantic_tokens_request);
}
```

trying to compile this I get the following error,
```
error[E0308]: mismatched types
   --> sway-lsp/src/server.rs:858:9
    |
858 |         test_lsp_cap(doc_comments_dir(), semantic_tokens_request);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
    |
    = note: expected trait `for<'r, 's> <for<'r, 's> fn(&'r mut LspService<server::Backend>, &'s tower_lsp::lsp_types::Url) -> impl Future<Output = tower_lsp::jsonrpc::Request> {semantic_tokens_request} as FnOnce<(&'r mut LspService<server::Backend>, &'s tower_lsp::lsp_types::Url)>>`
               found trait `for<'r, 's> <for<'r, 's> fn(&'r mut LspService<server::Backend>, &'s tower_lsp::lsp_types::Url) -> impl Future<Output = tower_lsp::jsonrpc::Request> {semantic_tokens_request} as FnOnce<(&'r mut LspService<server::Backend>, &'s tower_lsp::lsp_types::Url)>>`
note: the lifetime requirement is introduced here
   --> sway-lsp/src/server.rs:843:50
    |
843 |         F: Fn(&mut LspService<Backend>, &Url) -> Fut,
    |                                                  ^^^

For more information about this error, try `rustc --explain E0308`.
```